server/blob/blob-api/src/test/java/org/apache/james/blob/api/BucketBlobStoreContract.java

```diff
@@ -0,0 +1,109 @@
+
+public interface BucketBlobStoreContract {
+    String SHORT_STRING = "toto";
+    byte[] SHORT_BYTEARRAY = SHORT_STRING.getBytes(StandardCharsets.UTF_8);
+    BucketName CUSTOM = BucketName.of("custom");
+
+    int THREAD_COUNT = 10;
+    int OPERATION_COUNT = 10;
+
+    BlobStore testee();
+
+    BlobId.Factory blobIdFactory();
+
+    @Test
+    default void saveBytesShouldThrowWhenNullBucketName() {
+        assertThatThrownBy(() -> testee().save(null, SHORT_BYTEARRAY).block())
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void saveStringShouldThrowWhenNullBucketName() {
+        assertThatThrownBy(() -> testee().save(null, SHORT_STRING).block())
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void saveInputStreamShouldThrowWhenNullBucketName() {
+        assertThatThrownBy(() -> testee().save(null, new ByteArrayInputStream(SHORT_BYTEARRAY)).block())
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void readShouldThrowWhenNullBucketName() {
+        BlobId blobId = testee().save(BucketName.DEFAULT, SHORT_BYTEARRAY).block();
+        assertThatThrownBy(() -> testee().read(null, blobId))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void readBytesStreamShouldThrowWhenNullBucketName() {
+        BlobId blobId = testee().save(BucketName.DEFAULT, SHORT_BYTEARRAY).block();
+        assertThatThrownBy(() -> testee().readBytes(null, blobId).block())
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void readStringShouldThrowWhenBucketDoesNotExist() {
+        BlobId blobId = testee().save(BucketName.DEFAULT, SHORT_BYTEARRAY).block();
+        assertThatThrownBy(() -> testee().read(CUSTOM, blobId))
+            .isInstanceOf(ObjectStoreException.class);
+    }
+
+    @Test
+    default void readBytesStreamShouldThrowWhenBucketDoesNotExist() {
+        BlobId blobId = testee().save(BucketName.DEFAULT, SHORT_BYTEARRAY).block();
+        assertThatThrownBy(() -> testee().readBytes(CUSTOM, blobId).block())
+            .isInstanceOf(ObjectStoreException.class);
+    }
+
+    @Test
+    default void shouldBeAbleToSaveDataInMultipleBuckets() {
+        BlobId blobIdDefault = testee().save(BucketName.DEFAULT, SHORT_BYTEARRAY).block();
+        BlobId blobIdCustom = testee().save(CUSTOM, SHORT_BYTEARRAY).block();
+
+        byte[] bytesDefault = testee().readBytes(BucketName.DEFAULT, blobIdDefault).block();
+        byte[] bytesCustom = testee().readBytes(CUSTOM, blobIdCustom).block();
+
+        assertThat(bytesDefault).isEqualTo(bytesCustom);
+    }
+
+    @Test
+    default void saveConcurrentlyWithNonPreExistingBucketShouldNotFail() throws Exception {
+        ConcurrentTestRunner.builder()
+            .operation(((threadNumber, step) -> testee().save(CUSTOM, SHORT_STRING + threadNumber + step).block()))
+            .threadCount(THREAD_COUNT)
```

Comment:
I find it less readable with a constant than having the raw value here. It's the upside of having a builder: the name of the concept is the method name, so you don't need to put the "magic value" into a named variable or a constant.